crates/bevy_tasks/src/task_pool.rs

```diff
@@ -266,67 +270,166 @@ impl TaskPool {
-            f(scope_ref);
-
-            if spawned.is_empty() {
-                Vec::new()
-            } else {
+        Self::THREAD_EXECUTOR
+            .with(|thread_executor| self.scope_with_executor_inner(true, thread_executor, f))
+    }
+
+    /// This allows passing an external executor to spawn tasks on. When you pass an external executor
+    /// [`Scope::spawn_on_scope`] spawns is then run on the thread that [`ThreadExecutor`] is being ticked on.
+    /// If [`None`] is passed the scope will use a [`ThreadExecutor`] that is ticked on the current thread.
+    ///
+    /// When `tick_task_pool_executor` is set to `true`, the multithreaded task stealing executor is ticked on the scope
+    /// thread. Disabling this can be useful when finishing the scope is latency sensitive. Pulling tasks from
+    /// global excutor can run tasks unrelated to the scope and delay when the scope returns.
+    ///
+    /// See [`Self::scope`] for more details in general about how scopes work.
+    pub fn scope_with_executor<'env, F, T>(
+        &self,
+        tick_task_pool_executor: bool,
+        thread_executor: Option<&ThreadExecutor>,
+        f: F,
+    ) -> Vec<T>
+    where
+        F: for<'scope> FnOnce(&'scope Scope<'scope, 'env, T>),
+        T: Send + 'static,
+    {
+        // If a `thread_executor` is passed use that. Otherwise get the `thread_executor` stored
+        // in the `THREAD_EXECUTOR` thread local.
+        if let Some(thread_executor) = thread_executor {
+            self.scope_with_executor_inner(tick_task_pool_executor, thread_executor, f)
+        } else {
+            Self::THREAD_EXECUTOR.with(|thread_executor| {
+                self.scope_with_executor_inner(tick_task_pool_executor, thread_executor, f)
+            })
+        }
+    }
+
+    fn scope_with_executor_inner<'env, F, T>(
+        &self,
+        tick_task_pool_executor: bool,
+        thread_executor: &ThreadExecutor,
+        f: F,
+    ) -> Vec<T>
+    where
+        F: for<'scope> FnOnce(&'scope Scope<'scope, 'env, T>),
+        T: Send + 'static,
+    {
+        // SAFETY: This safety comment applies to all references transmuted to 'env.
+        // Any futures spawned with these references need to return before this function completes.
+        // This is guaranteed because we drive all the futures spawned onto the Scope
+        // to completion in this function. However, rust has no way of knowing this so we
+        // transmute the lifetimes to 'env here to appease the compiler as it is unable to validate safety.
+        let executor: &async_executor::Executor = &self.executor;
+        let executor: &'env async_executor::Executor = unsafe { mem::transmute(executor) };
+        let thread_executor: &'env ThreadExecutor<'env> =
+            unsafe { mem::transmute(thread_executor) };
+        let spawned: ConcurrentQueue<async_executor::Task<T>> = ConcurrentQueue::unbounded();
+        let spawned_ref: &'env ConcurrentQueue<FallibleTask<T>> =
```

Comment:
This type now changes during the transmute, which feels more than a little bit sus. Was this intentional? If so, can you explain the intent (and why it is ok?). FallibleTask does contain async_executor::Task internally and nothing else (currently), but that doesn't mean it always will, or that new values won't be added. This feels like a liability, even if it is technically valid currently (which I can't say with certainty). I thought you needed `repr(transparent)` for this, which FallibleTask does not use.

Reply:
This is a mistake. Nice catch. I must have messed up a merge. 

I'm actually surprised this didn't fail CI. I've made similar mistakes before and it usually crashes.